bcipy/language/model/gpt2.py

```diff
@@ -81,84 +182,122 @@ def __get_char_predictions(self, word_prefix: str) -> List[tuple]:
         evidence_str = evidence_str.replace(SPACE_CHAR, ' ')
-
-        # if we are at the start of a word, let the language model predict the next word
-        # then we get the probability distribution of the first character of the next word
-        if self.is_start_of_word is True:
-            self.curr_word_predicted_prob = self.__model_infer(evidence_str)
-            next_char_pred = self.__get_char_predictions("")
-
-        # if we are at the middle of a word, just get the marginal distribution of
-        # the next character given current word prefix
-        else:
-            # first we need the language model predict the probability distribution of the
-            # current word
-            self.curr_word_predicted_prob = self.__model_infer(" ".join(evidence_str.split()[:-1]))
-            cur_word_prefix = evidence_str.split()[-1]
-            next_char_pred = self.__get_char_predictions(cur_word_prefix)
+        evidence_str = evidence_str.lower()
+
+        # infer the first wordpiece
+        predicted_logit = self.__model_infer(evidence_str, True)
+        sorted_idx = predicted_logit.argsort()[::-1]
+
+        # get the indices of wordpieces whose have prefix match with the last wordpiece of the
+        # evidence_str
+        matched_sorted_idx, new_evidence_str = self.__prefix_match(sorted_idx, evidence_str)
+
+        all_beam_candidates = []
+
+        # construct the initial beam
+        beam = []
+        for i in range(min(beam_width, len(matched_sorted_idx))):
+            candidate = ([matched_sorted_idx[i]], predicted_logit[matched_sorted_idx[i]])
+            beam.append(candidate)
+
+        # populate all_beam_candidates with tuples of (text, log_likelihood)
+        self.__populate_beam_candidates(new_evidence_str, beam, all_beam_candidates)
+
+        # further beam search with more depth
+        for depth in range(search_depth - 1):
+            new_candidates = []
+            for candidate in beam:
+                candidate_tokens_idx = candidate[0]
+                candidate_tokens_log_prob = candidate[1]
+                candidate_text = new_evidence_str + self.tokenizer.decode(candidate_tokens_idx)
+                predicted_logit = self.__model_infer(candidate_text, False)
+                sorted_idx = predicted_logit.argsort()[::-1]
+                for i in range(beam_width):
+                    # make sure that wordpieces contain only letters
+                    if self.idx_to_word[sorted_idx[i]].isalpha():
+                        new_candidate = (candidate_tokens_idx + [sorted_idx[i]],
+                                         candidate_tokens_log_prob + predicted_logit[sorted_idx[i]])
+                        new_candidates.append(new_candidate)
+
+            # sort the new candidates based on likelihood and populate the beam
+            ordered_candidates = sorted(new_candidates, key=lambda x: x[1], reverse=True)
+            beam = ordered_candidates[:beam_width]
```

Comment:
It might be useful to factor out the beam search algorithm from the specifics of how we're using it. This should be generalizable.